### ISSUE TITLE
consistent-this: option to disallow this aliases completely

### ISSUE THREAD

**Issue by mourner** [eslint/eslint#1745]
It's often considered a bad practice to use `var self = this` trick for referencing outer context, and instead recommended to use `bind`. It would be awesome to have an option in the `consistent-this` rule to disable `this` aliases completely:

```js
// warning
var self = this;
button.onclick = function () {
  console.log(self);
};

// ok
button.onclick = function () {
  console.log(this);
}.bind(this);
```

**Comment by nzakas** [comment]
I'm not sure that's an appropriate option for `consistent-this`, as it has nothing to do with the current rule, but open to hearing arguments.

**Comment by mourner** [comment]
@nzakas makes sense; would it be appropriate to have a separate `no-this-aliases` instead?

**Comment by nzakas** [comment]
This is where it gets tricky. We try very hard not to have rules that are mutually exclusive. So a separate rule would clash directly with this rule if both were turned on at the same time (would be nonsensical to do so, but we've seen things like this happen before and people get frustrated). Need to ponder this a bit more.

**Comment by rlidwka** [comment]
> It's often considered a bad practice to use var self = this trick for referencing outer context, and instead recommended to use bind

What would you suggest to do if someone needs to reference both inner context `this` and outer context `this` at the same time?

**Comment by mourner** [comment]
@rlidwka do you have an example of a situation where this is needed?

**Comment by rlidwka** [comment]
Yeah, I remember a few times where callback's `this` value was needed, and it was used inside a class. Can't find an example right now.

**Comment by nzakas** [comment]
Since adding a new rule would conflict with consistent-this, I think the best path forward is for you to create a custom rule for this purpose.
